File: src/builtins/trait_constructor.rs
use crate::builtins::*;
use crate::fundamentals::*;

#[derive(Clone)]
pub struct TraitConstructor {
    pub id: AnyTraitID,
    pub validation: AnyValidation,
}

impl TraitConstructor {
    pub fn new<T: Clone>(id: TraitID<T>, validation: Validation<Value, T>) -> TraitConstructor {
        TraitConstructor {
            id: AnyTraitID::from(id),
            validation: AnyValidation::from(validation),
        }
    }
}

simple_trait! {
    name: trait_constructor,
    type: TraitConstructor,
    label: "Trait",
}

pub(crate) fn init(env: &mut Environment) {
    // Trait ::= Text
    env.add_conformance(Conformance::new(
        TraitID::text,
        TraitID::trait_constructor.validation(),
        |_, _| Ok(Text(String::from("<trait>"))),
    ))
}

File: Sources/Builtins/TraitConstructor.swift
public struct TraitConstructor {
    public var id: AnyTraitID
    public var validation: AnyValidation

    public init<T>(
        id: TraitID<T>,
        validation: @escaping AnyValidation
    ) {
        self.id = AnyTraitID(id)
        self.validation = validation
    }
}

extension TraitID where T == TraitConstructor {
    public static var traitConstructor: Self {
        .builtin("Trait")
    }
}

extension Trait where T == TraitConstructor {
    public static func traitConstructor(_ value: TraitConstructor) -> Self {
        Trait(id: .traitConstructor) { _, _ in value }
    }
}

internal func setupTraitConstructor(_ env: inout Environment) {
    // Trait ::= Text
    env.addConformance(
        Conformance(
            derivedTraitID: .text,
            validation: TraitID.traitConstructor.validation,
            deriveTraitValue: { _, _, _ in Text("<trait>") }
        )
    )
}
